### ISSUE TITLE
XML / XSL Lexer

### ISSUE THREAD

**Issue by ajtruckle** [ScintillaOrg/lexilla#243]
Hi, I am new to Scintilla and Lexilla. I can’t see a lever for XML or XSL files. Can someone please advise me on which lexer to use?

Thank you.

**Comment by ajtruckle** [comment]
@zufuliu 
Thank you. So the HTML lexer doubles up as a XML lexer. 

In which case I am confused because when I passed "xml" to create the lexer I ended up with:

1. No folder margin.
2. No syntax highlighting. 

I am going to try out the dedicated Scintilla Editor to see what it should look like.

**Comment by zufuliu** [comment]
> No syntax highlighting.

https://www.scintilla.org/ScintillaDoc.html#Lexer
https://github.com/ScintillaOrg/lexilla/blob/e793865c57daba24c49a97567efaa51cf056ddb9/lexers/LexHTML.cxx#L808-L810

> No folder margin.

https://github.com/ScintillaOrg/lexilla/blob/e793865c57daba24c49a97567efaa51cf056ddb9/lexers/LexHTML.cxx#L776-L780

**Comment by ajtruckle** [comment]
@zufuliu 
I never set the **fold.html** option to enable folding. Let me try again. 

Concerning **syntax highlighting** , I am not sure what you are saying.

**Comment by zufuliu** [comment]
You need to define styles (font, color, etc.) listed in `lexicalClassesHTML`, otherwise all HTML elements will be rendered with same style (`STYLE_DEFAULT`).

**Comment by ajtruckle** [comment]
@zufuliu 
Ah, ok. I assumed that the lexer would have an initial colour scheme configured as a footprint.

**Comment by zufuliu** [comment]
https://www.scintilla.org/ScintillaDoc.html#StyleDefinition
https://sourceforge.net/p/scintilla/scite/ci/default/tree/src/html.properties

**Comment by ajtruckle** [comment]
@zufuliu 

I am using the `CScintillaCtrl` MFC wrapper and I have now got 90% of the syntax highlighting working, based on the SciTE code you linked to.

This is what I have so far (for others benefit):

```cpp
if (strType == L"xml")
{
	SetAStyle(SCE_H_DEFAULT, RGB(0, 0, 0)); // $(font.text)
	SetAStyle(SCE_H_TAG, RGB(0, 0, 128));
	SetAStyle(SCE_H_TAGUNKNOWN, RGB(0, 0, 128));
	SetAStyle(SCE_H_ATTRIBUTE, RGB(0, 128, 128));
	SetAStyle(SCE_H_ATTRIBUTEUNKNOWN, RGB(0, 128, 128));
	// SCE_H_ERRORATTRIBUTEUNKNOWN
	// SetAStyle(SCE_H_NUMBER, RGB(0, 0, 0)); $(colour.number)
	// SetAStyle(SCE_H_DOUBLESTRING, RGB(0, 0, 0)); $(colour.string)
	// SetAStyle(SCE_H_SINGLESTRING, RGB(0, 0, 0)); $(colour.string)
	SetAStyle(SCE_H_OTHER, RGB(128, 0, 128));
	SetAStyle(SCE_H_COMMENT, RGB(128, 128, 0));
	SetAStyle(SCE_H_ENTITY, RGB(128, 0, 128)); // $(font.text)
	SetAStyle(SCE_H_TAGEND, RGB(0, 0, 128));
	SetAStyle(SCE_H_XMLSTART, RGB(128, 0, 128));
	m_edit.StyleSetBold(SCE_H_XMLSTART, true);
	SetAStyle(SCE_H_XMLEND, RGB(128, 0, 128));
	m_edit.StyleSetBold(SCE_H_XMLEND, 1);
	SetAStyle(SCE_H_CDATA, RGB(128, 0, 0), RGB(255, 240, 240)); // $(font.text)
	m_edit.StyleSetEOLFilled(SCE_H_CDATA, true);
	SetAStyle(SCE_H_QUESTION, RGB(128, 0, 0));
	SetAStyle(SCE_H_VALUE, RGB(96, 128, 96));
	SetAStyle(SCE_H_SGML_DEFAULT, RGB(0, 0, 128)); // $(colour.hypertext.sgml.back)
	SetAStyle(SCE_H_SGML_COMMAND, RGB(0, 0, 128)); // $(colour.hypertext.sgml.back)
	m_edit.StyleSetBold(SCE_H_SGML_COMMAND, 1); // $(colour.hypertext.sgml.back)
	SetAStyle(SCE_H_SGML_1ST_PARAM, RGB(0, 102, 0)); // $(colour.hypertext.sgml.back)
	SetAStyle(SCE_H_SGML_DOUBLESTRING, RGB(128, 0, 0)); // $(colour.hypertext.sgml.back)
	SetAStyle(SCE_H_SGML_SIMPLESTRING, RGB(153, 51, 0)); // $(colour.hypertext.sgml.back)
	SetAStyle(SCE_H_SGML_ERROR, RGB(51, 102, 255), RGB(255, 102, 102));
	SetAStyle(SCE_H_SGML_SPECIAL, RGB(0, 0, 0)); // $(colour.hypertext.sgml.back)
	SetAStyle(SCE_H_SGML_ENTITY, RGB(51, 51, 51)); // $(colour.hypertext.sgml.back)
	SetAStyle(SCE_H_SGML_COMMENT, RGB(128, 128, 0)); // $(colour.hypertext.sgml.back)
	SetAStyle(SCE_H_SGML_BLOCK_DEFAULT, RGB(204, 204, 224), RGB(204, 204, 224));
}
```

And this is what it looks like:

![image](https://github.com/ScintillaOrg/lexilla/assets/7985331/84d73f73-9cc2-4b23-8bec-f126e2da33a3)

And folders is working too! Although, I would like to have my lines in the margin to look like these here:

![image](https://github.com/ScintillaOrg/lexilla/assets/7985331/8d2b91ef-3e01-4aba-a992-d83fd47a2de5)

**Comment by ajtruckle** [comment]
@zufuliu 

Looks like I have to tweak the values I am using:

```cpp
	DefineMarker(static_cast<int>(Scintilla::MarkerOutline::FolderOpen), Scintilla::MarkerSymbol::CirclePlus, RGB(0xff, 0xff, 0xff), RGB(0, 0, 0xFF));
	DefineMarker(static_cast<int>(Scintilla::MarkerOutline::Folder), Scintilla::MarkerSymbol::CircleMinus, RGB(0xff, 0xff, 0xff), RGB(0, 0, 0));
	DefineMarker(static_cast<int>(Scintilla::MarkerOutline::FolderSub), Scintilla::MarkerSymbol::Empty, RGB(0xff, 0xff, 0xff), RGB(0, 0, 0));
	DefineMarker(static_cast<int>(Scintilla::MarkerOutline::FolderTail), Scintilla::MarkerSymbol::Empty, RGB(0xff, 0xff, 0xff), RGB(0, 0, 0));
	DefineMarker(static_cast<int>(Scintilla::MarkerOutline::FolderEnd), Scintilla::MarkerSymbol::Empty, RGB(0xff, 0xff, 0xff), RGB(0, 0, 0));
	DefineMarker(static_cast<int>(Scintilla::MarkerOutline::FolderOpenMid), Scintilla::MarkerSymbol::CirclePlusConnected, RGB(0xff, 0xff, 0xff), RGB(0, 0, 0));
	DefineMarker(static_cast<int>(Scintilla::MarkerOutline::FolderMidTail), Scintilla::MarkerSymbol::Empty, RGB(0xff, 0xff, 0xff), RGB(0, 0, 0));
```

Are these markers described anywhere?

**Comment by ajtruckle** [comment]
@zufuliu Progress with folding:

![image](https://github.com/ScintillaOrg/lexilla/assets/7985331/4da67741-d01c-448c-9646-5d452148fd66)

Code:

```cpp
DefineMarker(static_cast<int>(Scintilla::MarkerOutline::FolderOpen), Scintilla::MarkerSymbol::CircleMinus, RGB(0xff, 0xff, 0xff), RGB(0, 0, 0xFF));
DefineMarker(static_cast<int>(Scintilla::MarkerOutline::Folder), Scintilla::MarkerSymbol::CirclePlus, RGB(0xff, 0xff, 0xff), RGB(0, 0, 0));
DefineMarker(static_cast<int>(Scintilla::MarkerOutline::FolderSub), Scintilla::MarkerSymbol::VLine, RGB(0xff, 0xff, 0xff), RGB(0, 0, 0));
DefineMarker(static_cast<int>(Scintilla::MarkerOutline::FolderTail), Scintilla::MarkerSymbol::LCorner, RGB(0xff, 0xff, 0xff), RGB(0, 0, 0));
DefineMarker(static_cast<int>(Scintilla::MarkerOutline::FolderEnd), Scintilla::MarkerSymbol::CirclePlus, RGB(0xff, 0xff, 0xff), RGB(0, 0, 0));
DefineMarker(static_cast<int>(Scintilla::MarkerOutline::FolderOpenMid), Scintilla::MarkerSymbol::CircleMinus, RGB(0xff, 0xff, 0xff), RGB(0, 0, 0));
DefineMarker(static_cast<int>(Scintilla::MarkerOutline::FolderMidTail), Scintilla::MarkerSymbol::LCorner, RGB(0xff, 0xff, 0xff), RGB(0, 0, 0));
```

So I have two remaining questions before I can close this ticket:

1. In the above screen shot the final corner looks odd. Which property do I have that is set incorrectly? The rest appear right to me.
2. In the `SetAStyle` code, where I ported it from SciTE, I did not know how to port `$(colour.number)` etc. What are these property values?

Thank you for your help.
